atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImpl.java

```diff
@@ -28,32 +30,34 @@
 import com.palantir.logsafe.logger.SafeLoggerFactory;
 import java.util.Collection;
 import java.util.Comparator;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.NavigableMap;
+import java.util.Iterator;
 import java.util.Optional;
 import java.util.Set;
-import java.util.TreeMap;
-import java.util.stream.Collectors;
-import javax.annotation.concurrent.NotThreadSafe;
-
-@NotThreadSafe
+import java.util.SortedSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+@ThreadSafe
 final class SnapshotStoreImpl implements SnapshotStore {
     private static final SafeLogger log = SafeLoggerFactory.get(SnapshotStoreImpl.class);
 
-    private final NavigableMap<Sequence, ValueCacheSnapshot> snapshotMap;
-    private final Multimap<Sequence, StartTimestamp> liveSequences;
-    private final Map<StartTimestamp, Sequence> timestampMap;
+    private final ConcurrentNavigableMap<Sequence, ValueCacheSnapshot> snapshotMap;
+    private final SetMultimap<Sequence, StartTimestamp> liveSequences;
+    private final ConcurrentMap<StartTimestamp, Sequence> timestampMap;
     private final RateLimiter retentionRateLimiter = RateLimiter.create(1.0);
     private final int minimumUnusedSnapshots;
     private final int maximumSize;
 
     @VisibleForTesting
     SnapshotStoreImpl(int minimumUnusedSnapshots, int maximumSize, CacheMetrics cacheMetrics) {
-        this.snapshotMap = new TreeMap<>();
-        this.timestampMap = new HashMap<>();
-        this.liveSequences = MultimapBuilder.treeKeys().hashSetValues().build();
+        this.snapshotMap = new ConcurrentSkipListMap<>();
+        this.timestampMap = new ConcurrentHashMap<>();
+        this.liveSequences = Multimaps.synchronizedSetMultimap(
```

Comment:
may want to make this a `ConcurrentMap<Sequence, NavigableSet<StartTimestamp>>` similar to https://github.com/palantir/atlasdb/pull/6741 to avoid contention in hot paths